File: market-data/crypto-news/sdk/javascript/crypto-news.js
/**
 * Free Crypto News JavaScript SDK
 * 
 * 100% FREE - no API keys required!
 * Works in Node.js and browsers.
 * 
 * Usage:
 *   import { CryptoNews } from './crypto-news.js';
 *   const news = new CryptoNews();
 *   const articles = await news.getLatest(10);
 */

const DEFAULT_BASE_URL = 'https://cryptocurrency.cv';

export class CryptoNews {
  constructor(baseUrl = DEFAULT_BASE_URL) {
    this.baseUrl = baseUrl;
  }

  async _fetch(endpoint) {
    const response = await fetch(`${this.baseUrl}${endpoint}`);
    if (!response.ok) throw new Error(`HTTP ${response.status}`);
    return response.json();
  }

  /**
   * Get latest crypto news
   * @param {number} limit - Max articles (1-50)
   * @param {string} source - Filter by source
   * @returns {Promise<Array>} Articles
   */
  async getLatest(limit = 10, source = null) {
    let endpoint = `/api/news?limit=${limit}`;
    if (source) endpoint += `&source=${source}`;
    const data = await this._fetch(endpoint);
    return data.articles;
  }

  /**
   * Search news by keywords
   * @param {string} keywords - Comma-separated terms
   * @param {number} limit - Max results (1-30)
   */
  async search(keywords, limit = 10) {
    const encoded = encodeURIComponent(keywords);
    const data = await this._fetch(`/api/search?q=${encoded}&limit=${limit}`);
    return data.articles;
  }

  /** Get DeFi-specific news */
  async getDefi(limit = 10) {
    const data = await this._fetch(`/api/defi?limit=${limit}`);
    return data.articles;
  }

  /** Get Bitcoin-specific news */
  async getBitcoin(limit = 10) {
    const data = await this._fetch(`/api/bitcoin?limit=${limit}`);
    return data.articles;
  }

  /** Get breaking news (last 2 hours) */
  async getBreaking(limit = 5) {
    const data = await this._fetch(`/api/breaking?limit=${limit}`);
    return data.articles;
  }

  /** Get list of all sources */
  async getSources() {
    const data = await this._fetch('/api/sources');
    return data.sources;
  }

  /** Get trending topics */
  async getTrending(limit = 10, hours = 24) {
    return this._fetch(`/api/trending?limit=${limit}&hours=${hours}`);
  }

  /** Get API statistics */
  async getStats() {
    return this._fetch('/api/stats');
  }

  /** Check API health */
  async getHealth() {
    return this._fetch('/api/health');
  }

  /** Get news with topic classification and sentiment */
  async analyze(limit = 20, topic = null, sentiment = null) {
    let endpoint = `/api/analyze?limit=${limit}`;
    if (topic) endpoint += `&topic=${encodeURIComponent(topic)}`;
    if (sentiment) endpoint += `&sentiment=${sentiment}`;
    return this._fetch(endpoint);
  }

  /** Get archived news */
  async getArchive(date = null, query = null, limit = 50) {
    let endpoint = '/api/archive?';
    const params = [];
    if (date) params.push(`date=${date}`);
    if (query) params.push(`q=${encodeURIComponent(query)}`);
    params.push(`limit=${limit}`);
    return this._fetch(endpoint + params.join('&'));
  }

  /** Find original sources of news */
  async getOrigins(query = null, category = null, limit = 20) {
    let endpoint = '/api/origins?';
    const params = [`limit=${limit}`];
    if (query) params.push(`q=${encodeURIComponent(query)}`);
    if (category) params.push(`category=${category}`);
    return this._fetch(endpoint + params.join('&'));
  }

  /** Get portfolio news with optional prices */
  async getPortfolio(coins, limit = 10, includePrices = true) {
    const coinsParam = Array.isArray(coins) ? coins.join(',') : coins;
    return this._fetch(`/api/portfolio?coins=${encodeURIComponent(coinsParam)}&limit=${limit}&prices=${includePrices}`);
  }
}

// Convenience functions
export async function getCryptoNews(limit = 10) {
  return new CryptoNews().getLatest(limit);
}

export async function searchCryptoNews(keywords, limit = 10) {
  return new CryptoNews().search(keywords, limit);
}

// CommonJS compatibility
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { CryptoNews, getCryptoNews, searchCryptoNews };
}

File: market-data/crypto-news/sdk/python/crypto_news.py
"""
Free Crypto News Python SDK

100% FREE - no API keys required!

Usage:
    from crypto_news import CryptoNews
    
    news = CryptoNews()
    articles = news.get_latest(limit=10)
    
    for article in articles:
        print(f"{article['title']} - {article['source']}")
"""

import urllib.request
import json
from typing import Optional, List, Dict, Any

class CryptoNews:
    """Free Crypto News API client."""
    
    BASE_URL = "https://cryptocurrency.cv"
    
    def __init__(self, base_url: Optional[str] = None):
        """
        Initialize the client.
        
        Args:
            base_url: Optional custom API URL (for self-hosted instances)
        """
        self.base_url = base_url or self.BASE_URL
    
    def _request(self, endpoint: str) -> Dict[str, Any]:
        """Make API request."""
        url = f"{self.base_url}{endpoint}"
        with urllib.request.urlopen(url) as response:
            return json.loads(response.read().decode())
    
    def get_latest(self, limit: int = 10, source: Optional[str] = None) -> List[Dict]:
        """
        Get latest crypto news.
        
        Args:
            limit: Max articles (1-50)
            source: Filter by source (coindesk, theblock, decrypt, etc.)
        
        Returns:
            List of news articles
        """
        endpoint = f"/api/news?limit={limit}"
        if source:
            endpoint += f"&source={source}"
        return self._request(endpoint)["articles"]
    
    def search(self, keywords: str, limit: int = 10) -> List[Dict]:
        """
        Search news by keywords.
        
        Args:
            keywords: Comma-separated search terms
            limit: Max results (1-30)
        
        Returns:
            List of matching articles
        """
        encoded = urllib.parse.quote(keywords)
        return self._request(f"/api/search?q={encoded}&limit={limit}")["articles"]
    
    def get_defi(self, limit: int = 10) -> List[Dict]:
        """Get DeFi-specific news."""
        return self._request(f"/api/defi?limit={limit}")["articles"]
    
    def get_bitcoin(self, limit: int = 10) -> List[Dict]:
        """Get Bitcoin-specific news."""
        return self._request(f"/api/bitcoin?limit={limit}")["articles"]
    
    def get_breaking(self, limit: int = 5) -> List[Dict]:
        """Get breaking news (last 2 hours)."""
        return self._request(f"/api/breaking?limit={limit}")["articles"]
    
    def get_sources(self) -> List[Dict]:
        """Get list of all news sources."""
        return self._request("/api/sources")["sources"]
    
    def get_trending(self, limit: int = 10, hours: int = 24) -> Dict:
        """Get trending topics with sentiment."""
        return self._request(f"/api/trending?limit={limit}&hours={hours}")
    
    def get_stats(self) -> Dict:
        """Get API statistics and analytics."""
        return self._request("/api/stats")
    
    def get_health(self) -> Dict:
        """Check API health status."""
        return self._request("/api/health")
    
    def analyze(self, limit: int = 20, topic: Optional[str] = None, sentiment: Optional[str] = None) -> Dict:
        """Get news with topic classification and sentiment analysis."""
        endpoint = f"/api/analyze?limit={limit}"
        if topic:
            endpoint += f"&topic={urllib.parse.quote(topic)}"
        if sentiment:
            endpoint += f"&sentiment={sentiment}"
        return self._request(endpoint)
    
    def get_archive(self, date: Optional[str] = None, query: Optional[str] = None, limit: int = 50) -> Dict:
        """Get archived historical news."""
        params = [f"limit={limit}"]
        if date:
            params.append(f"date={date}")
        if query:
            params.append(f"q={urllib.parse.quote(query)}")
        return self._request(f"/api/archive?{'&'.join(params)}")
    
    def get_origins(self, query: Optional[str] = None, category: Optional[str] = None, limit: int = 20) -> Dict:
        """Find original sources of news."""
        params = [f"limit={limit}"]
        if query:
            params.append(f"q={urllib.parse.quote(query)}")
        if category:
            params.append(f"category={category}")
        return self._request(f"/api/origins?{'&'.join(params)}")
    
    def get_portfolio(self, coins: list, limit: int = 10, include_prices: bool = True) -> Dict:
        """Get portfolio news with optional prices from CoinGecko."""
        coins_param = ','.join(coins) if isinstance(coins, list) else coins
        return self._request(f"/api/portfolio?coins={urllib.parse.quote(coins_param)}&limit={limit}&prices={str(include_prices).lower()}")


# Convenience functions
def get_crypto_news(limit: int = 10) -> List[Dict]:
    """Quick function to get latest news."""
    return CryptoNews().get_latest(limit)

def search_crypto_news(keywords: str, limit: int = 10) -> List[Dict]:
    """Quick function to search news."""
    return CryptoNews().search(keywords, limit)

def get_trending_topics(limit: int = 10) -> List[Dict]:
    """Quick function to get trending topics."""
    return CryptoNews().get_trending(limit)["trending"]


if __name__ == "__main__":
    # Demo
    print("📰 Latest Crypto News\n" + "=" * 50)
    news = CryptoNews()
    for article in news.get_latest(5):
        print(f"\n📌 {article['title']}")
        print(f"   🔗 {article['link']}")
        print(f"   📰 {article['source']} • {article['timeAgo']}")
    
    print("\n\n📊 Trending Topics\n" + "=" * 50)
    trending = news.get_trending(5)
    for topic in trending["trending"]:
        emoji = "🟢" if topic["sentiment"] == "bullish" else "🔴" if topic["sentiment"] == "bearish" else "⚪"
        print(f"{emoji} {topic['topic']}: {topic['count']} mentions")
